src/document-definitions-validator.js

```diff
@@ -165,12 +166,18 @@ function validateDocDefinition(docType, docDefinition) {
     validationErrors.push('missing a "propertyValidators" property');
   }
 
-  function validatePermissions(permissionsCategory, permissionsDefinition) {
+  function validatePermissions(permissionsCategory, permissionsDefinition, supportedPermissionOperations) {
     var hasPermissionOperations = false;
     for (var permissionOperation in permissionsDefinition) {
       hasPermissionOperations = true;
       var permissions = permissionsDefinition[permissionOperation];
 
+      if (permissionOperation === 'replace' && (docDefinition.immutable === true || docDefinition.cannotReplace === true)) {
+        validationErrors.push('the "' + permissionsCategory + '" property\'s "' + permissionOperation + '" operation type is invalid when the document type is immutable');
+      } else if (permissionOperation === 'remove' && (docDefinition.immutable === true || docDefinition.cannotDelete === true)) {
+        validationErrors.push('the "' + permissionsCategory + '" property\'s "' + permissionOperation + '" operation type is invalid when the document type is immutable');
+      }
```

Comment:
It does not make sense to declare channels, roles and/or users for the `replace` and `remove` operations if the `immutable`, `cannotReplace` and/or `cannotDelete` constraints are variously enabled.